crates/kernel/src/device/sdcard.rs

```
@@ -0,0 +1,1789 @@
+            sderror!("do_data_command() called with buffer size less than block size")
+        }
+
+        self.emmc.blocks_to_transfer = (buf.len() / self.emmc.block_size as usize) as i32;
+        if buf.len() % self.emmc.block_size as usize != 0 {
+            sderror!(
+                "do_data_command() called with buffer size not an exact multiple of block size"
+            )
+        }
+
+        self.emmc.buf = buf.as_mut_ptr() as *mut ();
+        let command: u32;
+        if is_write {
+            if self.emmc.blocks_to_transfer > 1 {
+                command = WRITE_MULTIPLE_BLOCK;
+            } else {
+                command = WRITE_BLOCK;
+            }
+        } else {
+            if self.emmc.blocks_to_transfer > 1 {
+                command = READ_MULTIPLE_BLOCK;
+            } else {
+                command = READ_SINGLE_BLOCK;
+            }
+        }
+
+        let mut retry_count = 0;
+        let max_retries = 3;
+        while retry_count < max_retries {
+            if self.issue_command(command, block_no, 500000) {
+                break;
+            } else {
+                println!("Error sending CMD{command}");
+                println!("error = {:08x}", self.emmc.last_error);
+
+                retry_count += 1;
+                if retry_count < max_retries {
+                    println!("Retrying");
+                } else {
+                    println!("Giving up");
+                }
+            }
+        }
+
+        if retry_count == max_retries {
+            self.emmc.card_rca = CARD_RCA_INVALID;
+            return Err(SdCardError::Error);
+        }
+        Ok(())
+    }
+
+    fn do_read(&mut self, buf: &mut [u8], block_no: u32) -> Result<u32, SdCardError> {
+        if self.ensure_data_mode().is_err() {
+            return Err(SdCardError::Error);
+        }
+
+        if self.do_data_command(false, buf, block_no).is_err() {
+            return Err(SdCardError::Error);
+        }
```

These should propagate the same errors that caused it, rather than replacing them with `SdCardError::Error`;
there are three main ways of doing that, all of which are equivalent:
```rust
self.ensure_data_mode()?;
```
```rust
if let Err(e) = self.ensure_data_mode() {
    return Err(e);
}
```
```rust
match self.ensure_data_mode() {
    Ok(_) => (),
    Err(e) => return Err(e);
}
```
The first one is the most concise, but the others show what's actually being done (and if you need to convert the error, they can also make that more clear.